{% extends 'base.html' %}

{% block content %}

<h1>{{symbol}} –– {{ensembl}} </h1>

<div class="card">
  <div class="card-header">
    <b>HGNC symbol:</b> {{symbol}}
  </div>
  <div class="card-body">
    <!-- <h5 class="card-title">Special title treatment</h5> -->
    <p class="card-text"><b>Description:</b> {{description}}</p>
    <p class="card-text"><b>Type:</b> {{gene_biotype}}</p>
    <p class="card-text"><b>Genomic coordinate (hg38):</b> chr{{chromosome}}: {{start_position}} - {{end_position}}</p>
    <p class="card-text"><b>Uniprot: </b>
      <a href="https://www.uniprot.org/uniprot/{{uniprot}}" target="_blank">{{uniprot}}
        <i class="fa fa-external-link-alt" aria-hidden="true"></i>
      </a>

    </p>
  </div>
</div>

<br>

<h2 style="display: inline-block;">Co-expression modules</h2>

<i class="fa fa-question-circle" aria-hidden="true"
  data-toggle="tooltip" data-placement="right"
  data-html="true"
  title="
  <h6>Co-expression modules that contain the query gene</h6>
  <p>The co-expression modules are inferred using blockwise weighted gene correlation network analysis (WGCNA). The <b>supernetwork</b> indicates where the modules are found, colored by which tissue the query gene transcript comes from. The <b>stacked barplot</b> indicates the tissue composition of the identified modules.</p>
  <p>Information about co-expressed genes may help to ellucidate what function the query genes has in Coronary Artery Disease. Data on each module can be accessed by clicking the hyperlinks.</p>
  ">
</i>


<div class='row'>
  <div class='col-sm'>
    <div id="super_network"></div>
  </div>

  <div class='col-sm'>
    <div id='module_barplot' ></div>
  </div>
</div>

<div class='row'>
  <div id='module_links' class='col'></div>
</div>


<br>

<h2 style="display: inline-block;">Absolute expression</h2>

<i class="fa fa-question-circle" aria-hidden="true"
  data-toggle="tooltip" data-placement="right"
  data-html="true"
  title="
  <h6>Tissue comparison of counts per million (CPM) quantifying the overall levels of gene expression</h6>
  <p>In general, high expression results, such as co-expression of highly expressed genes, is more likely to be biologically meaningful.</p>
  ">
</i>

<!-- Boxplot to be populated by plotly.js -->
<div id='cpm_boxplot'></div>

<h2 style="display: inline-block;">Differential expression</h2>

<i class="fa fa-question-circle" aria-hidden="true"
  data-toggle="tooltip" data-placement="right"
  data-html="true"
  title="
  <h6>Case-control comparison for Coronary Artery Disease patients and healthy controls</h6>
  ">
</i>


<table id='deg_table' class='display'></table>

<br>

<h2 style="display: inline-block;">Genetic variants (eQTLs)</h2>

<i class="fa fa-question-circle" aria-hidden="true"
  data-toggle="tooltip" data-placement="right"
  data-html="true"
  title="
  <h6>Association between genetic variants and gene expression, also known as expression quantitative trait loci (eQTL)</h6>
  ">
</i>


<table id='eqtl_table' class='display'></table>

{% endblock %}


{% block script %}

<!-- Plotly.js -->
<!-- <script src="https://cdn.plot.ly/plotly-latest.min.js"></script> -->
<script src="{{ url_for('bower.static', filename='plotly.js/dist/plotly.min.js')}}"></script>


<!-- Use assets for minimized code -->
{% assets 'js_gene' %}
<!-- <script type="text/javascript" src="{{ ASSET_URL }}"></script> -->
{% endassets %}

<!-- Use source code for development -->
<script src="{{ url_for('static', filename='js/table.js') }}"></script>
<script src="{{ url_for('static', filename='js/gene.js') }}"></script>

<script src="{{ url_for('static', filename='js/supernetw.js') }}"></script>

<!-- Set variables for client-side javascript from flask using Jinja -->
<script type="text/javascript">

	// User input for page, passed by Jinja variables
	var input = {
		gene: '{{ensembl}}'
	};
</script>

{% endblock %}