{% extends 'base.html' %}

{% block content %}

<!-- Inline block allows  -->
<!-- <h1 style="display: inline-block;">Co-expression module {{ mod_id }}</h1> -->
<h1 style="display: inline-block;">Co-expression module {{ mod_id }}
	&nbsp (H<sup>2</sup>=<span id="netw_heritability"></span>%)
</h1>

<i class="fa fa-question-circle" aria-hidden="true"
	data-toggle="tooltip" data-placement="right"
	data-html="true"
	title="
	<h6>Co-expression module inferred using blockwise weighted gene correlation network analysis (WGCNA).</h6>
	<p>The <b>network heritability</b> H<sup>2</sup> estimate from significant eSNPs, excluding variants in linkage disequilibrium (LD>0.2) with known loci identified by GWAS.</p>
	<p>The <b>circle diagram</b> shows the tissue distribution of the transcripts in this module. Modules are inferred across tissues and can therefore contain genes from multiple tissues.</p>
	<p><b>Module associations</b> to clinical measures are calculated by aggregating gene-level association p-values using Fisher's method. Association with differential expressed genes (DEG) is estimated by enrichment of DEG signatures using a Hypergeometric test.
	</p>
	">
</i>

<!-- Two equal Bootstrap columns -->
<div class='row'>
	<div class='col-sm center-block'>
		<p>The co-expression module contains <b>n=<span id="mod_size"></span></b> genes from the following tissues:</p>
			<div id="tissue_pie" style="width: 300px; height: 290px;"></div>
		<!-- <p>The module transcripts in aggregated are associated with the following phenotypes:</p> -->
	</div>
	<div class='col-sm'>
		<div id="pheno_assoc"></div>
	</div>
</div>

<h2 style="display: inline-block;">Regulatory gene network</h2>

<i class="fa fa-question-circle" aria-hidden="true"
	data-toggle="tooltip" data-placement="right"
	data-html="true"
	title="
	<h6>Interactive visualization of GENIE3 network inferred over the module genes.</h6>
	<p><b>Nodes</b> are colored based on the transcript tissue.</p>
	">
</i>

<div id="rgn">
	<div id="rgn_slider" style="display: inline-block"></div>

	<button type="button" class="btn btn-light" onclick="exportNetworkSVG()">SVG</button>

	<a href="/static/data/rgn/edges/{{mod_id}}.csv">
		<button type="button" class="btn btn-light">CSV</button>
	</a>

	<!-- Annotation of network nodes -->
	Annotation:
	<select id="annot_opts">
		<!-- Default option -->
		<option value="Tissue">Tissue</option>  
	</select>

</div>

<br>

<h2 style="display: inline-block;">Key drivers</h2>

<i class="fa fa-question-circle" aria-hidden="true"
	data-toggle="tooltip" data-placement="right"
	data-html="true"
	title="
	<h6>From the GENIE3 network influential genes (key drivers) are detected using the Mergomics key driver analysis.</h6>
	">
</i>

<div id="kda">
	<table id='kda_table' class='display'></table>
</div>

<h2 style="display: inline-block;">Endocrine candidates</h2>

<i class="fa fa-question-circle" aria-hidden="true"
	data-toggle="tooltip" data-placement="right"
	data-html="true"
	title="
	<h6>Secreted genes from cross-tissue co-expression modules with significant correlation with tissue-specific eigengene.</h6>
	<p>The <b>supernetwork</b> column indacates whether (1) or not (0) the interaction between the from and target module is captured in the eigengene supernetwork.</p>
	">
</i>

<div id="endocrine_div">
	<table id="endocrine_table" class='display'></table>
</div>

<h2 style="display: inline-block">Genes</h2>

<i class="fa fa-question-circle" aria-hidden="true"
	data-toggle="tooltip" data-placement="right"
	data-html="true"
	title="
	<h6>Table of genes in the co-expression module.</h6>
	">
</i>

<table id='mod_table' class='display'></table>

<br>

<h2 style="display: inline-block;">Gene Ontology enrichment</h2>

<i class="fa fa-question-circle" aria-hidden="true"
	data-toggle="tooltip" data-placement="right"
	data-html="true"
	title="
	<h6>Top enrichment results for the module genes. Enrichment is estimated using the WGCNA method from the R package.</h6>
	">
</i>


<!-- Tab navigation for GO tables -->
<nav>
 	<div class="nav nav-tabs" id="nav-tab" role="tablist">
		<a class="nav-item nav-link active" id="nav-go_bp-tab" data-toggle="tab" href="#nav-go_bp" role="tab" aria-controls="nav-go_bp" aria-selected="true">Biological processes</a>
		<a class="nav-item nav-link" id="nav-go_cc-tab" data-toggle="tab" href="#nav-go_cc" role="tab" aria-controls="nav-go_cc" aria-selected="false">Cellular components</a>
		<a class="nav-item nav-link" id="nav-go_mf-tab" data-toggle="tab" href="#nav-go_mf" role="tab" aria-controls="nav-go_mf" aria-selected="false">Molecular function</a>
 	</div>
</nav>

<!-- Content (tables) for Gene Ontology enrichment -->
<div class="tab-content" id="nav-tabContent">
	<div class="tab-pane fade show active" id="nav-go_bp" role="tabpanel" aria-labelledby="nav-go_bp-tab">
		<table id="go_bp_table" class="display"></table>
	</div>
	<div class="tab-pane fade" id="nav-go_cc" role="tabpanel" aria-labelledby="nav-go_cc-tab">
		<table id="go_cc_table" class="display"></table>
	</div>
	<div class="tab-pane fade" id="nav-go_mf" role="tabpanel" aria-labelledby="nav-go_mf-tab">
		<table id="go_mf_table" class="display"></table>
	</div>
</div>

<br>

<h2 style="display: inline-block;">Module eQTL</h2>

<i class="fa fa-question-circle" aria-hidden="true"
  data-toggle="tooltip" data-placement="right"
  data-html="true"
  title="
  <h6>Association between genetic variants and gene expression, also known as expression quantitative trait loci (eQTL)</h6>
  <p>Only shows the single, most significant eQTL per gene.</p>
  ">
</i>


<table id='eqtl_table' class='display'></table>


{% endblock %}

{% block script %}

<!-- Plotly.js -->
<!-- <script src="https://cdn.plot.ly/plotly-latest.min.js"></script> -->
<script src="{{ url_for('bower.static', filename='plotly.js/dist/plotly.min.js')}}"></script>

<script src="{{ url_for('static', filename='js/table.js') }}"></script>
<script src="{{ url_for('static', filename='js/module.js') }}"></script>

<script type="text/javascript">

	// User input for page, passed by Jinja variables
	var input = {
		mod_id: '{{mod_id}}'
	};
</script>

{% endblock %}
